### PR TITLE
correctly return the max version from searches

### PR DESCRIPTION
**Pull request by mwrock**
Our internal cookbooks get reved on each push so the build numbers can grow rapidly. We are having issues where the string comparison on version breaks and returns an incorrect version. I think the submitted test illustrates this. Given two cookbook versions, 1.0.24 and 1.0.8, berks currently returns 1.0.8.

This PR uses `Semverse::Version` to do the comparison so that the versions are compared more accurately.

### CODE REVIEW

File: lib/berkshelf/source.rb [hunk 1]
```diff
@@ -53,7 +53,7 @@ def search(name)
       universe
         .select { |cookbook| cookbook.name =~ Regexp.new(name) }
         .group_by(&:name)
-        .collect { |name, versions| versions.max_by(&:version) }
+        .collect { |name, versions| versions.max_by { |v| Semverse::Version.new(v.version) } }
```
Comment (sethvargo) [inline lib/berkshelf/source.rb]: Hrm, sooooo this is a _really_ expensive operation. Parsing a semver version was one of the larger bottlenecks in our previous Ruby-based depsolver.

I think, if we do this, we will need to implement some comparisons using bit-shifts in semverse to make this feasible. With hundreds of cookbooks, you could be approaching a minute to sort :bullettrain_side:

**Comment by mwrock** [comment]
Thanks @sethvargo thats really good to know and obviously unacceptable in terms of perf. I hadn't looked into the Semverse comparison code or done any load testing. I'll look more closely at this. Its currently become a blocker for us and I'm wiring this into our CI now so will be curious to see perf ramifications.

**Comment by mwrock** [comment]
yup. did some basic bench marking passing 100,000 random versions doing a string compare vs. `Semverse::Version` and the later is consistently 4x slower to find the max. no surprise really. Out of curiosity I ran the same through Rubygems::Version and it was ~ 3x slower and a small tweak optimizing it for versions that are mostly numeric (a fair assumption i think) was a bit over 2x slower.

So I'm gonna do some playing later on with some bit-foo to try to optimize further.

That all said, having run the PR as is through our chef server of hundreds of CB versions for the past few days has been quite acceptable, but I don't think it will be too bad to get this much faster.

**Comment by mwrock** [comment]
Here is an optimized version. Its pretty ugly but attempts to leverage some micro optimizations. Its still slower than a straight up string compare but considerably (~3x) faster than `Semverse::Version` or `Rubygem::Version`. I've kept it separate from `Semverse::Version` to avoid the split in the initializer. The comparison assumes that it is dealing with valid version strings since it is only used to consume persisted versions.

I'll admit there may be a better approach here. I'm fairly new to ruby and there are likely some advanced optimizations I'm missing or some logic holes I've missed because my brain cell count is is in a perpetual decline. I've tried to capture most edge cases in the unit tests.

**Comment by mwrock** [comment]
This could actually be ported to `Semverse::Version` without perf penalty if we make the splits lazy. I'd be happy to do that but will hold off until I get a thumbs up.

**Comment by sethvargo** [comment]
@mwrock I would rather do something like bit-shifting here. For example:

```ruby
parts = version.split(/[\.\+]/)
(0..5).each do |i|
  parts[i] ||= "3"
end
parts.map! { |i| i.to_i }

parts[0] << 32 | parts[1] << 16 | parts[2] << 8 | parts[3] << 4 | parts[4] << 2 | parts[5]
```

You'll need to account for `.beta` and stuff as well, but that will convert any "reasonable" version string to a sortable integer. It will fail if any "segment" is over a 32-bits though /cc @reset.

**Comment by mwrock** [comment]
Interesting. I may be doing something incorrectly here but bench marking that reveals it to be even more expensive than `Semverse::Version` and for some reason Ruby-Prof throws a `stack level too deep` when I turn it on for that version. Here is my hacked up implementation of `<=>`:

```
    def <=>(other)
      return 0 if @version == other.version

      my_parts = @version.split(/[\.\+]/)
      (0..5).each do |i|
        my_parts[i] ||= "3"
      end
      my_parts.map! { |i| i.to_i }

      other_parts = other.version.split(/[\.\+]/)
      (0..5).each do |i|
        other_parts[i] ||= "3"
      end
      other_parts.map! { |i| i.to_i }

      b = 32

      for i in 0..my_parts.length-1
        return 1 if other_parts[i].nil?
        mp = my_parts[i] << b
        op = other_parts[i] << b

        # puts "#{mp}::#{op}"
        return mp < op ? -1 : 1 unless mp == op
        b = b / 2
      end
      return -1
    end
```

I threw this up in a bit of a rush so another set of eyes may uncover something obvious.

The tricky thing with these versions is handling prereleases and thus the ability to evaluate both as string and int. My PR implementation basically operates on the initial version stream as a byte stream and attempts to traverse as few bytes as possible. The idea being that unless you have two adjacent version, the code flow nver touches the end of the version.

**Comment by mwrock** [comment]
Duh. I was missing the bitwise OR. However adding that really didnt impact perf.

Just to unblock rub-prof I commented out the actual part comparisons and returned -1 always. Here is the ruby-prof output that really just covers the version parsing:

```
 %self      total      self      wait     child     calls  name
 26.02      0.408     0.408     0.000     0.000   200000   String#split
 24.17      0.600     0.379     0.000     0.221   200000   Array#map!
 16.96      1.505     0.266     0.000     1.239   100000   Test::Version5#<=>
 14.09      0.221     0.221     0.000     0.000   900000   String#to_i
 11.73      0.231     0.184     0.000     0.047   100000   Range#each
  4.02      1.568     0.063     0.000     1.505        1   Array#each
  3.00      0.047     0.047     0.000     0.000   300000   Array#[]=
  0.00      1.568     0.000     0.000     1.568        1   Enumerable#max
  0.00      1.568     0.000     0.000     1.568        1   <Class::Test::Test>#yo
```

Considerable time is spent in split, map and casting which is along the lines of what I expected since these involve alot of object creation and enumerating. I'm trying to subvert that by to adding any objects and only enumerating once and hopefully not a full enumeration.

The nice thing with a straight up string compare like you have now is it doesn't really have to do any parsing at all and will always be about as fast as you can get. Unfortunately its also plain incorrect for version comparisons and that is likely a core problem for berkshelf since its one of Berkshelf's primary responsibilities (at least as I see it).

Anyhoo. I apologize for the comment spam. This has obviously triggered my OCD this weekend.

**Comment by reset** [comment]
@sethvargo you're on the right path but @mwrock is right, object creation and cleanup is where all the time is going to be spent. If we're really worried about performance we would want to drop down into C.

**Comment by mwrock** [comment]
@reset that makes sense to keep things logically together. I'm cool with whatever implementation works for the common berkshelf use case, My main concern is that version ordering results are what one would expect. The current string sort results in a bunch of missing cookbook errors against our internal berkshelf server and I'd love to not have to maintain a fork as we upgrade chefDKs in the future. I'd be happy to contribute the current optimized ugliness, the original `Semverse::Version` PR or something in between.

We pull cookbooks from an internal chef server backed berkshelf api service and fall back to the supermarket. The Semverse tweak has not yet produced uncomfortable latency, but I'm not sure how "typical" our environment is compared to others.

This was a fun way to become better familiar with ruby internals, profiling and optimization techniques. Not much of a surprise that they aren't that different from languages I'm better acquainted with.

Thanks @reset and @sethvargo for having a look.

**Comment by reset** [comment]
Even though this contains potential perf issues, I think we should accept the Pr because it does contain a valid fix.

We can work on a more performant implementation if we begin to see issues in the wild.

/cc @sethvargo

Jamie Winsor
@resetexistence

**Comment by mwrock** [comment]
I just reset the PR branch removing the optimizations. So now its just a vanilla call to `Semverse::Version`

**Comment by sethvargo** [comment]
kk fair enough :+1:

**Comment by sonots** [comment]
@sethvargo I am waiting this fix will be released. do you have any plans on releasing new version?

**Comment by repeatedly** [comment]
I also need new version including this fix.
This is very critical problem.
